torch/_prims_common/__init__.py

```diff
@@ -844,22 +844,36 @@ def infer_size(shape: ShapeType, numel: int) -> Tuple[int, ...]:
             newsize *= d
         else:
             torch._check(False, lambda: f"invalid shape dimension {d}")
-    torch._check(
-        numel == newsize or (dim is not None and newsize > 0 and numel % newsize == 0),
-        lambda: f"shape '{list(shape)}' is invalid for input of size {numel}",
-    )
-    if dim is not None:
-        # Convert to list to produce a compatible error message with core
-        # PyTorch, which prints sequences in square brackets.
-        shape = list(shape)
+    if dim is None:
+        torch._check(numel == newsize)
+    else:
         torch._check(
             newsize != 0,
```

Comment:
this check is made redundant with the new one below.

Reply:
I know, but this error message is better :)